Give a warning before uninstalling the active/default toolchain, or the last/host target for a toolchain

### Problem you are trying to solve

Rustup currently lets you do some silly things:
```
; rustup target list --installed
x86_64-unknown-linux-gnu
; rustup target remove x86_64-unknown-linux-gnu
info: removing component 'rust-std'
; cargo build
   Compiling inner v0.1.0 (/home/jyn/src/example/inner)
error[E0463]: can't find crate for `std`
```
This isn't a bug, but it seems ... confusing.

### Solution you'd like

 I think rustup should warn in either of the following cases:
1. If you're trying to remove the host target, which means you can no longer build proc-macros or build scripts
2. If you're trying to remove the last remaining target installed
3. If you're uninstalling the active/default toolchain

### Notes

_No response_
```[tasklist]
### Tasks
- [ ] https://github.com/rust-lang/rustup/pull/3637
- [ ] https://github.com/rust-lang/rustup/pull/3924
```

--- a/src/cli/rustup_mode.rs
+++ b/src/cli/rustup_mode.rs
@@ -1281,8 +1281,25 @@ async fn toolchain_link(
 }
 
 fn toolchain_remove(cfg: &mut Cfg<'_>, opts: UninstallOpts) -> Result<utils::ExitCode> {
+    let default_toolchain = cfg.get_default().ok().flatten();
+    let active_toolchain = cfg.find_active_toolchain().ok().flatten().map(|(it, _)| it);
+
     for toolchain_name in &opts.toolchain {
         let toolchain_name = toolchain_name.resolve(&cfg.get_default_host_triple()?)?;
+
+        if active_toolchain
+            .as_ref()
+            .is_some_and(|n| n == &toolchain_name)
+        {
+            warn!("removing the active toolchain; a toolchain override will be required for running Rust tools");
+        }
+        if default_toolchain
+            .as_ref()
+            .is_some_and(|n| n == &toolchain_name)
+        {
+            warn!("removing the default toolchain; proc-macros and build scripts might no longer build");
+        }
+
         Toolchain::ensure_removed(cfg, (&toolchain_name).into())?;
     }
     Ok(utils::ExitCode(0))

--- a/src/cli/rustup_mode.rs
+++ b/src/cli/rustup_mode.rs
@@ -1165,7 +1165,7 @@ async fn target_remove(
         let target = TargetTriple::new(target);
         let default_target = cfg.get_default_host_triple()?;
         if target == default_target {
-            warn!("after removing the default host target, proc-macros and build scripts might no longer build");
+            warn!("removing the default host target; proc-macros and build scripts might no longer build");
         }
         // Whether we have at most 1 component target that is not `None` (wildcard).
         let has_at_most_one_target = distributable
@@ -1179,7 +1179,7 @@ async fn target_remove(
             .at_most_one()
             .is_ok();
         if has_at_most_one_target {
-            warn!("after removing the last target, no build targets will be available");
+            warn!("removing the last target; no build targets will be available");
         }
         let new_component = Component::new("rust-std".to_string(), Some(target), false);
         distributable.remove_component(new_component).await?;

--- a/src/toolchain/names.rs
+++ b/src/toolchain/names.rs
@@ -359,6 +359,15 @@ from_variant!(
     LocalToolchainName::Path
 );
 
+impl PartialEq<ToolchainName> for LocalToolchainName {
+    fn eq(&self, other: &ToolchainName) -> bool {
+        match self {
+            LocalToolchainName::Named(n) => n == other,
+            _ => false,
+        }
+    }
+}
+
 impl Display for LocalToolchainName {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {

--- a/tests/suite/cli_v2.rs
+++ b/tests/suite/cli_v2.rs
@@ -212,15 +212,36 @@ async fn list_toolchains_with_none() {
 }
 
 #[tokio::test]
-async fn remove_toolchain() {
+async fn remove_toolchain_default() {
     let mut cx = CliTestContext::new(Scenario::SimpleV2).await;
     cx.config.expect_ok(&["rustup", "update", "nightly"]).await;
     cx.config
-        .expect_ok(&["rustup", "toolchain", "remove", "nightly"])
+        .expect_stderr_ok(
+            &["rustup", "toolchain", "remove", "nightly"],
+            "removing the default toolchain; proc-macros and build scripts might no longer build",
+        )
         .await;
     cx.config.expect_ok(&["rustup", "toolchain", "list"]).await;
     cx.config
         .expect_stdout_ok(&["rustup", "toolchain", "list"], "no installed toolchains")
+        .await;
+}
+
+#[tokio::test]
+async fn remove_toolchain_active() {
+    let mut cx = CliTestContext::new(Scenario::SimpleV2).await;
+    cx.config.expect_ok(&["rustup", "default", "nightly"]).await;
+    cx.config
+        .expect_ok(&["rustup", "override", "set", "stable"])
+        .await;
+    cx.config
+        .expect_stderr_ok(
+            &["rustup", "toolchain", "remove", "stable"],
+            "removing the active toolchain; a toolchain override will be required for running Rust tools",
+        )
+        .await;
+    cx.config
+        .expect_stdout_ok(&["rustup", "toolchain", "list"], "nightly")
         .await;
 }
 

--- a/tests/suite/cli_v2.rs
+++ b/tests/suite/cli_v2.rs
@@ -1208,10 +1208,12 @@ async fn remove_target_host() {
     cx.config
         .expect_ok(&["rustup", "target", "add", clitools::CROSS_ARCH1])
         .await;
-    cx.config.expect_stderr_ok(
-        &["rustup", "target", "remove", &host],
-        "after removing the default host target, proc-macros and build scripts might no longer build",
-    ).await;
+    cx.config
+        .expect_stderr_ok(
+            &["rustup", "target", "remove", &host],
+            "removing the default host target; proc-macros and build scripts might no longer build",
+        )
+        .await;
     let path = format!("toolchains/nightly-{host}/lib/rustlib/{host}/lib/libstd.rlib");
     assert!(!cx.config.rustupdir.has(path));
     let path = format!("toolchains/nightly-{host}/lib/rustlib/{host}/lib");
@@ -1228,7 +1230,7 @@ async fn remove_target_last() {
     cx.config
         .expect_stderr_ok(
             &["rustup", "target", "remove", &host],
-            "after removing the last target, no build targets will be available",
+            "removing the last target; no build targets will be available",
         )
         .await;
 }

